src/rgw/rgw_appmain.cc

```diff
@@ -587,28 +603,34 @@ void rgw::AppMain::shutdown(std::function<void(void)> finalize_async_signals)
     fe->stop();
   }
 
-  for (auto& fe : fes) {
-    fe->join();
-    delete fe;
-  }
-
-  for (auto& fec : fe_configs) {
-    delete fec;
-  }
-
   ldh.reset(nullptr); // deletes
   finalize_async_signals(); // callback
   rgw_log_usage_finalize();
-  
+
```

Comment:
this changed the ordering between `finalize_async_signals()` and `fe->join();`

`fe->stop();` above calls `RGWLibFrontend::stop()` -> `RGWProcessFrontend::stop()` -> `thread->kill(SIGUSR1)`

that signal was racing with `finalize_async_signals()` from `RGWLib::stop()` to remove that signal handler

that's why librgw shutdown hits `FAILED ceph_assert(handlers[signum])`